pkg/cortex/server_service.go

```diff
@@ -0,0 +1,55 @@
+package cortex
+
+import (
+	"context"
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/weaveworks/common/server"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+// NewServerService constructs service from Server component.
+// servicesToWaitFor is called when server is stopping, and should return all
+// services that need to terminate before server actually stops.
+func NewServerService(serv *server.Server, servicesToWaitFor func() []services.Service) services.Service {
+	serverDone := make(chan error, 1)
+
+	runFn := func(ctx context.Context) error {
+		go func() {
+			defer close(serverDone)
+			serverDone <- serv.Run()
+		}()
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-serverDone:
+			if err != nil {
+				level.Info(util.Logger).Log("msg", "server failed", "err", err)
+			}
+			return err
+		}
+	}
+
+	stoppingFn := func() error {
+		// wait until all modules are done, and then shutdown server.
+		for _, s := range servicesToWaitFor() {
+			_ = s.AwaitTerminated(context.Background())
+		}
+
+		// unblock Run, if it's still running (e.g. service was asked to stop via StopAsync)
+		serv.Stop()
+
+		// shutdown HTTP and gRPC servers
+		serv.Shutdown()
```

Comment:
Is it safe to call `Shutdown()` if `Run()` is already running?

Reply:
Yes. Shutdown will stop HTTP and gRPC servers, which makes Run stop. I've added some basic tests for server service.

Reply:
I've removed `serv.Stop` from this method and use it in the test instead (can't be used in both places, as it would double-close channel, which panics).